src/Features/Core/Portable/SplitOrMergeIfStatements/Consecutive/AbstractMergeConsecutiveIfStatementsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,146 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.Formatting;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.SplitOrMergeIfStatements
+{
+    internal abstract class AbstractMergeConsecutiveIfStatementsCodeRefactoringProvider<TExpressionSyntax>
+        : AbstractMergeIfStatementsCodeRefactoringProvider<TExpressionSyntax>
+        where TExpressionSyntax : SyntaxNode
+    {
+        protected abstract bool IsElseClauseOfIfStatement(SyntaxNode node, out SyntaxNode ifStatementNode);
+
+        protected abstract bool HasElseClauses(SyntaxNode ifStatementNode);
+
+        protected abstract SyntaxNode MergeIfStatements(
+            SyntaxNode firstIfStatementNode, SyntaxNode secondIfStatementNode, TExpressionSyntax condition);
+
+        protected sealed override CodeAction CreateCodeAction(Func<CancellationToken, Task<Document>> createChangedDocument)
+            => new MyCodeAction(createChangedDocument, IfKeywordText);
+
+        protected sealed override async Task<bool> CanBeMergedAsync(
+            Document document, SyntaxNode ifStatement, ISyntaxFactsService syntaxFacts, CancellationToken cancellationToken)
+        {
+            return CanBeMergedWithParent(syntaxFacts, ifStatement) ||
+                   await CanBeMergedWithPreviousStatementAsync(document, syntaxFacts, ifStatement, cancellationToken).ConfigureAwait(false);
+        }
+
+        protected sealed override SyntaxNode GetChangedRoot(
+            SyntaxNode root, SyntaxNode ifStatement, ISyntaxFactsService syntaxFacts, SyntaxGenerator generator)
+        {
+            var previousIfStatement = IsElseClauseOfIfStatement(ifStatement, out var parentIfStatement)
+                ? parentIfStatement
+                : GetPreviousStatement(syntaxFacts, ifStatement);
+
+            var newCondition = (TExpressionSyntax)generator.LogicalOrExpression(
+                syntaxFacts.GetIfStatementCondition(previousIfStatement),
+                syntaxFacts.GetIfStatementCondition(ifStatement));
```

Comment:
is this a new method?  SyntaxFacts methods are usuallly named "GetXXXOfYYY".  This is to prevent hte problem of having methods like "GetInvocationExpressionExpression".  Instead, it's "GetExpressionOfInvocationExpression".

Reply:
Yes, it is a new method. Will rename.

Reply:
Since the method also returns the condition of an else-if, would you prefer it to be named `GetConditionOfIfLike`? or `GetConditionOfIfOrElseIf`?

Reply:
➡️ The method was moved out of `ISyntaxFactsService` into `IIfLikeStatementGenerator`.